pkg/controller/compute/address.go

```diff
@@ -0,0 +1,156 @@
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package compute
+
+import (
+	"context"
+	"time"
+
+	"github.com/crossplane/provider-gcp/pkg/clients/address"
+
+	"github.com/google/go-cmp/cmp"
+	"google.golang.org/api/compute/v1"
+	"k8s.io/client-go/util/workqueue"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/controller"
+
+	xpv1 "github.com/crossplane/crossplane-runtime/apis/common/v1"
+	"github.com/crossplane/crossplane-runtime/pkg/errors"
+	"github.com/crossplane/crossplane-runtime/pkg/event"
+	"github.com/crossplane/crossplane-runtime/pkg/logging"
+	"github.com/crossplane/crossplane-runtime/pkg/meta"
+	"github.com/crossplane/crossplane-runtime/pkg/ratelimiter"
+	"github.com/crossplane/crossplane-runtime/pkg/reconciler/managed"
+	"github.com/crossplane/crossplane-runtime/pkg/resource"
+
+	"github.com/crossplane/provider-gcp/apis/compute/v1beta1"
+	gcp "github.com/crossplane/provider-gcp/pkg/clients"
+)
+
+// Error strings.
+const (
+	errNotAddress = "managed resource is not an Address"
+)
+
+// SetupAddress adds a controller that reconciles Address managed resources.
+func SetupAddress(mgr ctrl.Manager, l logging.Logger, rl workqueue.RateLimiter, poll time.Duration) error {
+	name := managed.ControllerName(v1beta1.AddressGroupKind)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		WithOptions(controller.Options{
+			RateLimiter: ratelimiter.NewDefaultManagedRateLimiter(rl),
+		}).
+		For(&v1beta1.Address{}).
+		Complete(managed.NewReconciler(mgr,
+			resource.ManagedKind(v1beta1.AddressGroupVersionKind),
+			managed.WithExternalConnecter(&addressConnector{kube: mgr.GetClient()}),
+			managed.WithReferenceResolver(managed.NewAPISimpleReferenceResolver(mgr.GetClient())),
+			managed.WithConnectionPublishers(),
+			managed.WithPollInterval(poll),
+			managed.WithLogger(l.WithValues("controller", name)),
+			managed.WithRecorder(event.NewAPIRecorder(mgr.GetEventRecorderFor(name)))))
+}
+
+type addressConnector struct {
+	kube client.Client
+}
+
+func (c *addressConnector) Connect(ctx context.Context, mg resource.Managed) (managed.ExternalClient, error) {
+	projectID, opts, err := gcp.GetAuthInfo(ctx, c.kube, mg)
+	if err != nil {
+		return nil, err
+	}
+	s, err := compute.NewService(ctx, opts)
+	if err != nil {
+		return nil, errors.Wrap(err, errNewClient)
+	}
+	return &addressExternal{kube: c.kube, Service: s, projectID: projectID}, errors.Wrap(err, errNewClient)
+}
+
+type addressExternal struct {
+	kube      client.Client
+	projectID string
+	*compute.Service
+}
+
+func (e *addressExternal) Observe(ctx context.Context, mg resource.Managed) (managed.ExternalObservation, error) {
+	cr, ok := mg.(*v1beta1.Address)
+	if !ok {
+		return managed.ExternalObservation{}, errors.New(errNotAddress)
+	}
+
+	region := gcp.StringValue(cr.Spec.ForProvider.Region)
```

Comment:
I am wondering whether the `region` parameter in the spec of the `Address` resource needs to be enforced? Or whether one needs to return an error here (and the related functions) if the region is not specified. The Google docs are a bit unclear on this IMO.

Reply:
Hmm, experimenting with the Go GCP SDK, it seems the omission of the region parameter transparently creates a "global" address w/o throwing an error. This means the `globalAddresses` REST endpoint seems pretty much useless. One can get the same behaviour by setting or omitting the region in the `addresses` endpoint. Of course this does not mean that Crossplane could at this point not enforce the region parameter and this way force the user to either use `GlobalAddress` or `Adress`. As it stands now, `Address` could be used to create global as well as regional addresses.

Reply:
Actually, I was wrong, the _region_ parameter is required. Without it, one gets the following error from the underlying SDK / REST API: 

```
 Error 400: Invalid value for field 'region': ''. Must be a match of regex '[a-z](?:[-a-z0-9]{0,61}[a-z0-9])?', invalid"}
```

The question is what to do here? Should the region be explicitly checked or is it enough to rely on the underlying error of the SDK which seems to be quite descriptive? The rest of the codebase seems to not do any explicit checking either, so it might be good enough to rely on the SDK error. 
Yet another approach would be a validating webhook. This would be nice since it would prevent even applying an "invalid" CR. On the other hand, I don't see validating webhook setup yet. Maybe that would make another good new feature - adding a validating webhook. WDYT?

Reply:
I would recommend you to use `region` as a required parameter. The whole point of creating a new managed resource (`Address`) is to give the user the ability to create an `Address` with `region` parameter, because the `GlobalAddress` didn't have it. 

As you pointed out, this is a required parameter for the API (and the SDK, because it is a parameter for the CRUD methods), so I think it makes sense to put it as required.

Reply:
I'll add `+kubebuilder:validation:Required` to the CRD spec.

Reply:
Fixed, _region_ is now required.